File: app/src/main/java/rod/bailey/trafficatnsw/cameras/details/TrafficCameraImageActivity.java
package rod.bailey.trafficatnsw.cameras.details;

import android.app.ActionBar;
import android.app.Activity;
import android.app.AlertDialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Color;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.v4.app.NavUtils;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.widget.LinearLayout;
import android.widget.RelativeLayout.LayoutParams;

import java.io.InputStream;

import rod.bailey.trafficatnsw.R;
import rod.bailey.trafficatnsw.cameras.FavouriteCameraDialogPresenter;
import rod.bailey.trafficatnsw.cameras.TrafficCamera;
import rod.bailey.trafficatnsw.cameras.TrafficCameraDatabase;
import rod.bailey.trafficatnsw.util.MLog;

import static android.graphics.BitmapFactory.*;

public class TrafficCameraImageActivity extends Activity {

	private static final String TAG = TrafficCameraImageActivity.class
			.getSimpleName();
	private int cameraIndex;
	private String cameraStreet;
	private String cameraSuburb;
	private String cameraDescription;
	private String cameraUrl;

	private class DownloadImageTask extends AsyncTask<String, Void, Bitmap> {
		private final Context ctx;
		private ProgressDialog dialog;

		public DownloadImageTask(Context ctx) {
			assert ctx != null;
			this.ctx = ctx;
		}

		@Override
		protected void onPreExecute() {
			super.onPreExecute();

			Bitmap frameBitmap = decodeResource(getResources(),
					R.drawable.photo_frame_filled);

			mainLayout.removeAllViews();
			captionedImage = new CaptionedImageComponent(
					TrafficCameraImageActivity.this, frameBitmap,
					(Bitmap) null, cameraDescription);
			mainLayout.addView(captionedImage, captionedImageLLP);

			dialog = new ProgressDialog(ctx);
			dialog.setMessage("Loading image...");
			dialog.setCancelable(false);
			dialog.setIndeterminate(true);
			dialog.show();
		}

		protected Bitmap doInBackground(String... urls) {
			assert urls != null;
			assert urls.length == 1;

			MLog.i(TAG, "Up to doInBackground");

			String urlToLoad = urls[0];
			Bitmap result = null;

			try {
				InputStream stream = new java.net.URL(urlToLoad).openStream();
				result = BitmapFactory.decodeStream(stream);
			} catch (Exception e) {
				MLog.e(TAG, e.getMessage());
			}

			return result;
		}

		@Override
		protected void onPostExecute(Bitmap result) {
			if (dialog != null) {
				dialog.dismiss();
			}

			if (result == null) {
				// Failed to load camera image - bad connection or just not
				// available on server.
				MLog.i(TAG,
						"Failed to load camera image - showing error dialog");
				AlertDialog.Builder builder = new AlertDialog.Builder(ctx);
				builder.setTitle("Could not load camera image");
				builder.setMessage("Tap the refresh icon at top right to try again.");
				builder.setPositiveButton("OK", null);
				AlertDialog dialog = builder.create();
				dialog.show();
			} else {
				Bitmap frameBitmap = decodeResource(getResources(),
						R.drawable.photo_frame_with_black_corners);
				mainLayout.removeAllViews();
				captionedImage = new CaptionedImageComponent(
						TrafficCameraImageActivity.this, frameBitmap, result,
						cameraDescription);
				mainLayout.addView(captionedImage, captionedImageLLP);
			}
		}
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);

		Bundle extras = getIntent().getExtras();
		String strIndex = extras.getString("index");
		String strStreet = extras.getString("street");
		String strSuburb = extras.getString("suburb");
		String strDescription = extras.getString("description");
		String strUrl = extras.getString("url");
		String strFavourite = extras.getString("favourite");

		MLog.d(TAG, "index=" + strIndex);
		MLog.d(TAG, "stree=" + strStreet);
		MLog.d(TAG, "suburb=" + strSuburb);
		MLog.d(TAG, "description=" + strDescription);
		MLog.d(TAG, "url=" + strUrl);
		MLog.d(TAG, "favourite=" + strFavourite);

		assert strIndex != null;
		assert strStreet != null;
		assert strSuburb != null;
		assert strDescription != null;
		assert strUrl != null;
		assert strFavourite != null;

		cameraIndex = Integer.parseInt(strIndex);
		cameraStreet = strStreet.trim();
		cameraSuburb = strSuburb.trim();
		cameraDescription = strDescription.trim();
		cameraUrl = strUrl.trim();
		cameraFavourite = Boolean.parseBoolean(strFavourite);

		createUI();
	}

	// private final int IMAGE_VIEW_ID = 1000;
	//
	// private ImageView imageView;
	private MenuItem favouriteMenuItem;
	private boolean cameraFavourite;
	// private ImageView photoFrameImageView;
	private LinearLayout mainLayout;
	private CaptionedImageComponent captionedImage;
	private LinearLayout.LayoutParams captionedImageLLP;

	private void createUI() {
		mainLayout = new LinearLayout(this);
		mainLayout.setOrientation(LinearLayout.VERTICAL);
		mainLayout.setKeepScreenOn(true);
		mainLayout.setBackgroundColor(Color.BLACK);
		mainLayout.setGravity(Gravity.CENTER);

		captionedImage = new CaptionedImageComponent(this);
		captionedImageLLP = new LinearLayout.LayoutParams(
				LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
		mainLayout.addView(captionedImage, captionedImageLLP);

		setContentView(mainLayout);

		// Put camera title in action bar
		ActionBar actionBar = getActionBar();
		actionBar.setDisplayShowCustomEnabled(true);
		actionBar.setDisplayHomeAsUpEnabled(true);

		TrafficCameraTitleView customView = new TrafficCameraTitleView(this,
				cameraStreet, cameraSuburb);
		ActionBar.LayoutParams customViewALP = new ActionBar.LayoutParams(
				LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT);
		actionBar.setCustomView(customView, customViewALP);

		refresh();
	}

	public void refresh() {
		MLog.i(TAG, "Refreshing");
		DownloadImageTask task = new DownloadImageTask(this);
		task.execute(cameraUrl);
	}

	public void toggleFavourite() {
		MLog.i(TAG, "Toggle favourite");

		// Raise dialog
		final TrafficCamera camera = TrafficCameraDatabase.getInstance()
				.getCamera(cameraIndex);
		FavouriteCameraDialogPresenter pres = new FavouriteCameraDialogPresenter(
				camera);

		AlertDialog dialog = pres.build(this);
		dialog.setOnDismissListener(new DialogInterface.OnDismissListener() {

			@Override
			public void onDismiss(DialogInterface dialog) {
				MLog.i(TAG,
						"On dismiss listener, camera.isFavourite="
								+ camera.isFavourite());
				updateActionBarToReflectFavouriteStatus(camera.isFavourite());
			}
		});
		dialog.show();

		MLog.i(TAG, "After dialog.show");
	}

	private void updateActionBarToReflectFavouriteStatus(boolean isFavourite) {
		favouriteMenuItem.setIcon(isFavourite ? R.drawable.ic_action_important
				: R.drawable.ic_action_not_important);
	}

	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		MenuInflater inflater = new MenuInflater(this);
		inflater.inflate(R.menu.traffic_camera_image_options_menu, menu);
		favouriteMenuItem = menu.findItem(R.id.toggle_camera_favourite);

		MLog.i(TAG, "Found favouriteMenuItem=" + favouriteMenuItem);
		updateActionBarToReflectFavouriteStatus(cameraFavourite);

		return super.onCreateOptionsMenu(menu);
	}

	@Override
	public boolean onOptionsItemSelected(MenuItem item) {
		if (item != null) {
			switch (item.getItemId()) {
			case R.id.refresh_camera_image:
				refresh();
				break;
			case R.id.toggle_camera_favourite:
				toggleFavourite();
				break;
			case android.R.id.home:
				NavUtils.navigateUpFromSameTask(this);
				break;
			}
		}
		return true;
	}
}

File: app/src/main/java/rod/bailey/trafficatnsw/cameras/details/TrafficCameraImageActivity.kt
package rod.bailey.trafficatnsw.cameras.details

import android.content.Context
import android.graphics.Bitmap
import android.os.Bundle
import android.support.v4.app.NavUtils
import android.support.v7.app.AppCompatActivity
import android.support.v7.widget.AppCompatTextView
import android.view.Menu
import android.view.MenuInflater
import android.view.MenuItem
import android.widget.ImageView
import org.androidannotations.annotations.AfterViews
import org.androidannotations.annotations.EActivity
import org.androidannotations.annotations.Extra
import org.androidannotations.annotations.ViewById
import rod.bailey.trafficatnsw.R
import rod.bailey.trafficatnsw.cameras.*

/**
 * Screen containing a single Card that has a traffic camera imageView at top
 * and a description of the imageView underneath. Image may be refreshed or
 * marked as a favourite by the user.
 */
@EActivity(R.layout.activity_camera_image)
open class TrafficCameraImageActivity : AppCompatActivity(), ITrafficCameraImageDisplayer {

	@ViewById(R.id.iv_camera_image)
	@JvmField
	var imageView: ImageView? = null

	@ViewById(R.id.tv_camera_title)
	@JvmField
	var titleTextView: AppCompatTextView? = null

	@ViewById(R.id.tv_camera_subtitle)
	@JvmField
	var subtitleTextView: AppCompatTextView? = null

	@ViewById(R.id.tv_camera_description)
	@JvmField
	var descriptionTextView: AppCompatTextView? = null

	@Extra("index")
	@JvmField
	var cameraIndex: Int? = null

	@Extra("street")
	@JvmField
	var cameraStreet: String? = null

	@Extra("suburb")
	@JvmField
	var cameraSuburb: String? = null

	@Extra("description")
	@JvmField
	var cameraDescription: String? = null

	@Extra("url")
	@JvmField
	var cameraUrl: String? = null

	@Extra("favourite")
	@JvmField
	var favourite: Boolean? = null

	private var favouriteMenuItem: MenuItem? = null
	private var cameraFavourite: Boolean = false

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		overridePendingTransition(R.anim.slide_in_from_right, R.anim.slide_out_to_left)
	}

	override fun displayImage(image: Bitmap) {
		imageView?.setImageBitmap(image)
	}

	@AfterViews
	fun afterViews() {
		val actionBar = supportActionBar
		actionBar?.title = getString(R.string.camera_image_screen_title)
		actionBar?.setDisplayShowCustomEnabled(true)
		actionBar?.setDisplayHomeAsUpEnabled(true)

		titleTextView?.text = cameraStreet
		subtitleTextView?.text = cameraSuburb
		descriptionTextView?.text = cameraDescription

		refresh()
	}

	fun refresh() {
		DownloadImageTask(context = this, displayer = this).execute(cameraUrl)
	}

	fun toggleFavourite() {
		val camera: TrafficCamera? = TrafficCameraCacheSingleton.instance.getCamera(cameraIndex ?: 0)
		if (camera != null) {
			val pres = FavouriteCameraDialogPresenter(camera)
			val dialog = pres.build(this)
			dialog.setOnDismissListener {
				updateActionBarPerFavouriteStatus(camera.isFavourite)
			}
			dialog.show()
		}
	}

	private fun updateActionBarPerFavouriteStatus(isFavourite: Boolean) {
		favouriteMenuItem?.setIcon(
			if (isFavourite)
				R.drawable.ic_action_favourite_dark
			else
				R.drawable.ic_action_not_favourite_dark)
	}

	override fun onCreateOptionsMenu(menu: Menu): Boolean {
		val inflater = menuInflater
		inflater.inflate(R.menu.menu_traffic_camera_image_options, menu)
		favouriteMenuItem = menu.findItem(R.id.toggle_camera_favourite)
		updateActionBarPerFavouriteStatus(cameraFavourite)
		return true
	}

	override fun onOptionsItemSelected(item: MenuItem?): Boolean {
		if (item != null) {
			when (item.itemId) {
				R.id.refresh_camera_image -> refresh()
				R.id.toggle_camera_favourite -> toggleFavourite()
				android.R.id.home -> NavUtils.navigateUpFromSameTask(this)
			}
		}
		return true
	}

	companion object {
		fun start(ctx: Context, camera: TrafficCamera) {
			TrafficCameraImageActivity_.intent(ctx)
				.cameraIndex(camera.index)
				.cameraStreet(camera.street)
				.cameraSuburb(camera.suburb)
				.cameraDescription(camera.description)
				.cameraUrl(camera.url)
				.favourite(camera.isFavourite)
				.start()
		}
	}
}
